### PR TITLE
fix: save flow enabled for client credentials grant type

### LINKED ISSUE #31207: [Bug]: Authenticated API with OAuth 2.0 and Grant type "client credentials" is broken
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

When I want to use an Authenticated API data source with OAuth 2.0 authentication and the "client credentials" grant type, clicking on "Save & Authorize" fails with 500 and the following exception is logged:

```
appsmith  | backend stdout | java.lang.IllegalArgumentException: URI must not be null                                                                                                                             
appsmith  | backend stdout |    at org.springframework.util.Assert.notNull(Assert.java:204)                                                                                                                       
appsmith  | backend stdout |    at org.springframework.web.util.UriComponentsBuilder.fromUriString(UriComponentsBuilder.java:228)                                                                                 
appsmith  | backend stdout |    at com.appsmith.server.solutions.ce.AuthenticationServiceCEImpl.lambda$getAuthorizationCodeURLForGenericOAuth2$3(AuthenticationServiceCEImpl.java:145)                            
appsmith  | backend stdout |    at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:132)                                                                                                    
appsmith  | backend stdout |    at com.appsmith.server.configurations.MDCConfig$MdcContextLifter.onNext(MDCConfig.java:53)                                                                                        
appsmith  | backend stdout |    at reactor.core.publisher.MonoZip$ZipCoordinator.signal(MonoZip.java:293) 
appsmith  | backend stdout |    at reactor.core.publisher.MonoZip$ZipInner.onNext(MonoZip.java:474)
appsmith  | backend stdout |    at com.appsmith.server.configurations.MDCConfig$MdcContextLifter.onNext(MDCConfig.java:53)
...
```

By examining the code in `AuthenticationServiceCEImpl.java` I would suspect that the "client credentials" grant is not implemented correctly or is treated as an "authorization code" grant internally, as it requires an "Authorization URL". This is the cause of the `IllegalArgumentException: URI must not be null` from the stack trace above.

### Steps To Reproduce

The exception happens before any communication is done with the API, so no real OAuth 2 service is required to reproduce the failure, so `https://auth.example.com` is used as a placeholder.

1. Create an new application
2. Create a new "Authenticated API" datasource
3. Use `https://auth.example.com` as URL
4. Select "OAuth 2.0" as authentication
5. Select "Client Credentials" as grant type
6. Use `https://auth.example.com/oauth/token` or any other URL as "Access Token URL"
7. Insert `test` for "Client ID" and "Client Secret"
8. Use any scope like `openid`
9. Use `https://auth.example.com/` as "Audience" and "Resource"
10. Click on "Save & Authorize"
11. Receive a 500er response:

```
{
  "responseMeta": {
    "status": 500,
    "success": false,
    "error": {
      "code": "AE-APP-5000",
      "title": "Internal server error",
      "message": "Internal server error while processing request",
      "errorType": "INTERNAL_ERROR"
    }
  },
  "errorDisplay": "AE-APP-5000: Internal server error while processing request"
}
```

### Public Sample App

_No response_

### Environment

Production

### Issue video log

_No response_

### Version

Community Self Hosted - v1.13

### PR DESCRIPTION
**Pull request by NilanshBansal**
Fixes #31207 

## Automation

/ok-to-test tags="@tag.All"

### :mag: Cypress test results

> [!IMPORTANT]  
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/8423946100>
> Commit: `d19714118dda656f40ef5158a398d61ab39cef22`
> Cypress dashboard url: <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=8423946100&attempt=2" target="_blank">Click here!</a>
> All cypress tests have passed 🎉🎉🎉

## Summary by CodeRabbit

- **New Features**
    - Enhanced OAuth2 authorization support, allowing for different grant types.
    - Improved user interface for OAuth2 authorization, including specific messages under certain conditions.

### CODE REVIEW

**Review by sneha122 — APPROVED** [review]

**Comment by sneha122** [comment]
Tested the changes for following different DB types:
- Postgres (Databases)
- Authenticated and GraphQL API (REST)
- Gsheets (SaaS)

Looks good but found 2 issues,1 critical and 1 quality issue:
- After authorising gsheet datasource, I don't see the schema and preview:

https://github.com/appsmithorg/appsmith/assets/30018882/5a4d78ae-795c-430e-a516-79e89d7f4593

- When switching between auth code and client credentials, the height of the bottom tab is fluctuating

https://github.com/appsmithorg/appsmith/assets/30018882/382ddd43-90af-40de-a50a-2b38eb6b7f14

File: app/client/src/pages/common/datasourceAuth/index.tsx [hunk 1]
```diff
@@ -192,7 +202,10 @@ function DatasourceAuth({
   );
 
   useEffect(() => {
-    if (authType === AuthType.OAUTH2) {
+    if (
+      authType === AuthType.OAUTH2 &&
+      authorizationCodeGrantType === GrantType.AuthorizationCode
```
Comment (AmanAgarwal041) [inline app/client/src/pages/common/datasourceAuth/index.tsx]: Can we add a comment here to understand why this line was added ? As it narrows down the if condition. cc @NilanshBansal

Comment (NilanshBansal) [reply app/client/src/pages/common/datasourceAuth/index.tsx]: @AmanAgarwal041 can you review now. Have updated the code.

**Comment by AmanAgarwal041** [comment]
Found an issue with the API datasource form form:
https://github.com/appsmithorg/appsmith/assets/7565635/3a7cfbec-577c-4a94-bb6d-f9ab2c387517
cc @NilanshBansal

**Comment by AmanAgarwal041** [comment]
Other than the issue https://github.com/appsmithorg/appsmith/pull/32047#issuecomment-2018515166 , the made changes are working great.

**Comment by NilanshBansal** [comment]
> Found an issue with the API datasource form form:
> 
> [ API datasource form fields misbehaving ](https://www.loom.com/share/2c11a756899c4ec6b049186a4a50602f) [  ](https://www.loom.com/share/2c11a756899c4ec6b049186a4a50602f)
> cc @NilanshBansal

@AmanAgarwal041 this same behaviour is also reproducible in production env.
Can you please log a separate issue for this? The current PR only changes the Save button for Client Credentials grant type. 

https://github.com/appsmithorg/appsmith/assets/25542733/1785f700-edd9-4211-8f7c-17728a16f270

**Comment by AmanAgarwal041** [comment]
Logged a separate issue for the : https://github.com/appsmithorg/appsmith/pull/32047#issuecomment-2018515166

https://github.com/appsmithorg/appsmith/issues/32055

**Review by AmanAgarwal041 — APPROVED** [review]

**Pull request merged**